### PR TITLE
Fixes #149 : Simplify header escaping.

### LINKED ISSUE #149: escaping of SMTP folded subject lines introduces unexpected characters
According to the SMTP spec, long subject lines can be "folded", introducing carriage return and linefeed characters which are escaped improperly when retrieved via IMAP.

### PR DESCRIPTION
**Pull request by palimpsestor**

### CODE REVIEW

**Comment by buildscientist** [comment]
@palimpsestor - thanks for the PR. 

Is there any reason why you're removing the JavaEscape classes? Additionally I'm unsure if your escapeHeaders method covers all use cases.

**Comment by palimpsestor** [comment]
@buildscientist , the new tests included in ImapSubjectLineTest.java show cases which failed because of the usage of JavaEscape.  In some of those tests, as well as in the real usage of greenmail that prompted this PR, long subject lines can be "folded" by SMTP servers, introducing CLRF characters which are returned incorrectly as 'r' and 'n' when escaped with JavaEscape.

The admittedly simplified escapeHeaders method solves this problem, as well as satisfying all exisiting tests in the project, including those in EscapingTest.java.  If there are use cases which are not covered, we could add tests for those and adjust accordingly.

**Comment by buildscientist** [comment]
@palimpsestor Read up on folding in the RFC. I don't see anything in there that mentions \t characters are valid in a folded subject header - section 2.2.3 of the RFC states: 

> 
"The general rule is that wherever this standard allows for folding white space (not simply WSP characters), a CRLF may be inserted before any WSP."

File: greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java [hunk 1]
```diff
@@ -737,4 +736,9 @@ public static Header create(String[] header) {
             return new Header(header[0]);
         }
     }
+
+    public String escapeHeader(final String text) {
+        return text.replace("\r", "\\\r").replace("\n", "\\\n").replace("\\", "\\\\").replace("\"", "\\\"");
```
Comment (marcelmay) [inline greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java]: As discussed, let's use existing JavaMail mime utility for unfoldung and additionally escaping the quote.
As MimeUtility.quote(...) optionally surrounds the result with quotations, you'll also need to remove the quotation when using escapeHeader above.

    return MimeUtility.quote(MimeUtility.unfold(text), javax.mail.internet.HeaderTokenizer.RFC822);

File: greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java [hunk 2]
```diff
@@ -290,7 +289,7 @@ String parseEnvelope() {
         response.add(LB + Q + sentDateEnvelopeString + Q + SP);
         //2. Subject ---------------
         if (subject != null && (subject.length() != 0)) {
-            response.add(Q + JavaEscape.escapeJava(subject) + Q + SP);
+            response.add(Q + escapeHeader(subject) + Q + SP);
```
Comment (marcelmay) [inline greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java]: See comment for escapeHeader: Do not surround with quotation Q

File: greenmail-core/src/test/java/com/icegreen/greenmail/test/commands/ImapSubjectLineTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        testSubject(subjectWithDoubleQuote);
+    }
+
+    @Test
+    public void testSubjectWithTabCharacter() throws Exception {
+        String subjectWithTabCharacter = "The tab\t was there.";
+
+        testSubject(subjectWithTabCharacter);
+    }
+
```
Comment (marcelmay) [inline greenmail-core/src/test/java/com/icegreen/greenmail/test/commands/ImapSubjectLineTest.java]: Also test backslash: 

     @Test
    public void testSubjectWithBackSlashCharacter() throws Exception {
        String subjectWithTabCharacter = "With \\back slash.";

        testSubject(subjectWithTabCharacter);
    }

File: greenmail-core/src/test/java/com/icegreen/greenmail/test/commands/ImapSubjectLineTest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class ImapSubjectLineTest {
+
+    @Rule
+    public final GreenMailRule greenMail = new GreenMailRule(ServerSetupTest.ALL);
```
Comment (marcelmay) [inline greenmail-core/src/test/java/com/icegreen/greenmail/test/commands/ImapSubjectLineTest.java]: Better:

    ... = new GreenMailRule(ServerSetupTest.IMAP);

File: greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java [hunk 5]
```diff
@@ -737,4 +736,9 @@ public static Header create(String[] header) {
             return new Header(header[0]);
         }
     }
+
+    public String escapeHeader(final String text) {
```
Comment (marcelmay) [inline greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java]: private instead of public?

**Comment by marcelmay** [comment]
@palimpsestor : I finished with my comments on the pull diff. Your patch fixes the bug, reduces complexity and adds additional tests - great!

File: greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java [hunk 6]
```diff
@@ -381,7 +380,7 @@ String parseEnvelope() {
         }
         response.add(SP);
         if (messageID != null && messageID.length > 0) {
-            messageID[0] = JavaEscape.escapeJava(messageID[0]);
+            messageID[0] = escapeHeader(messageID[0]);
             response.add(Q + messageID[0] + Q);
```
Comment (marcelmay) [inline greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java]: See comment for escapeHeader: Obsolete surrounding with quotation Q  if additionally using MimeUtility.quote

File: greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java (hunk 1 above)
Comment (palimpsestor) [inline greenmail-core/src/main/java/com/icegreen/greenmail/store/SimpleMessageAttributes.java]: Using MimeUtility.unfold for the CLRFs worked great, but replacing the existing quotations with MimeUtility.quote (which only optionally quotes) caused several test failures.

**Comment by palimpsestor** [comment]
@marcelmay : Thanks for the feedback.  All good suggestions.  I only saw a bit ago that you had already merged this; I had incorporated all of those suggestions except for using MimeUtility for quoting, which was problematic (see inline comment).

**Comment by marcelmay** [comment]
@palimpsestor I saw that MimeUtility.quote caused some test failures. If you have any updates, reopen or open a new pull referencing the issue. We plan 1.5.1 on Sunday. Thanks alot!

**Pull request merged**